BeeSwift/AppDelegate.swift

```diff
@@ -142,4 +126,135 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UNUserNotificationCenterD
             completionHandler([.alert, .sound, .badge])
         }
     }
+    
+    // processing user's response to a notification
+    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
+        
+        guard response.actionIdentifier == UNNotificationDefaultActionIdentifier else {
+            completionHandler()
+            return
+        }
+        
+        guard let slug = response.notification.request.getSlug() else {
+            completionHandler()
+            return
+        }
+        
+        print("found slug: \(slug)")
+        
+        self.fetchGoalsIfNeededThenViewMatchingGoalWith(slug: slug)
+        
+        completionHandler()
+    }
+    
+    // MARK: opening goals
+    
+    private func fetchGoalsIfNeededThenViewMatchingGoalWith(slug: String) {
+        if CurrentUserManager.sharedManager.goals.isEmpty || CurrentUserManager.sharedManager.goalsFetchedAt.timeIntervalSinceNow < -3600 {
+            // apparently no data or stale data
+            
+            CurrentUserManager.sharedManager.fetchGoals(success: { goals in
+                CurrentUserManager.sharedManager.goals = goals
+                self.openGoalWith(slug)
+            }, error: nil)
+            
+        } else {
+            self.openGoalWith(slug)
+        }
+    }
+    
+    /// given a goal we shall navigate to it
+    private func openGoal(goal: JSONGoal) {
+        guard let navigationController = window?.rootViewController as? UINavigationController else { return }
+        
+        let alreadyShowingGoal = { () -> Bool in
+            guard let topVC = navigationController.topViewController as? GoalViewController else { return false }
+            return topVC.goal.slug == goal.slug
+        }()
+        
+        if !alreadyShowingGoal {
+            let goalVC: GoalViewController = {
+                let goalVC = GoalViewController()
+                goalVC.goal = goal
+                return goalVC
+            }()
+            navigationController.popToRootViewController(animated: true)
+            navigationController.pushViewController(goalVC, animated: true)
+        }
+    }
+    
+    /// for viewing, open goal with slug
+    /// - Parameter slug: slug of the goal to open
+    private func openGoalWith(_ slug: String) {
+        guard let goal = CurrentUserManager.sharedManager.goals.last(where: { $0.slug == slug }) else {
+            print("no goal matching slug: \(slug)")
+            
+            guard let navigationController = window?.rootViewController as? UINavigationController else { return }
+            navigationController.popToRootViewController(animated: true)
+            
+            return
+        }
+        self.openGoal(goal: goal)
+    }
+
+}
+
+// MARK: - private extensions for extracting slugs
+
+private extension UNNotificationRequest {
```

Comment:
private extensions were convenient but perhaps a _`SlugExtractor`_ helper would be better